File: prime-number/scala/src/main/scala/pn.scala
import scala.util.control.Breaks

class PrimeNumberInt {
  val primeNumbers = new scala.collection.mutable.ArrayBuffer[Int]()
  val b = new Breaks

  def isPrimeNumber(n: Int): Boolean = {
    var isPrime = true
    b.breakable {
      for (pn <- primeNumbers) {
        if (pn * pn > n) {
          isPrime = true
          b.break
        }
        if (n % pn == 0) {
          isPrime = false
          b.break
        }
      }
    }
    return isPrime
  }

  def calc(upperLimit: Int): Unit = {
    println("upper limit: " + upperLimit)
    for (n <- 2 to upperLimit) {
      if (isPrimeNumber(n))
        primeNumbers += n
    }
    println(s"Count: ${primeNumbers.length}")
  }
}

class PrimeNumberLong {
  val primeNumbers = new scala.collection.mutable.ArrayBuffer[Long]()
  val b = new Breaks

  def isPrimeNumber(n: Long): Boolean = {
    var isPrime = true
    b.breakable {
      for (pn <- primeNumbers) {
        if (pn * pn > n) {
          isPrime = true
          b.break
        }
        if (n % pn == 0) {
          isPrime = false
          b.break
        }
      }
    }
    return isPrime
  }

  def calc(upperLimit: Int): Unit = {
    println("upper limit: " + upperLimit)
    for (n <- 2 to upperLimit) {
      if (isPrimeNumber(n))
        primeNumbers += n
    }
    println(s"Count: ${primeNumbers.length}")
  }
}

class ArgParser(args: Array[String]) {
  var upperLimit = 0;
  var varTypeLong = false;
  val it = args.iterator
  while (it.hasNext) {
    it.next match {
      case "-u" => upperLimit = it.next.toInt
      case "--long" => varTypeLong = true
    }
  }
}

object Main {
  def main(args: Array[String]) = {
    val parser = new ArgParser(args)
    if (parser.varTypeLong) {
      println("VarType: Long")
      (new PrimeNumberLong()).calc(parser.upperLimit)
    } else {
      println("VarType: Int")
      (new PrimeNumberInt()).calc(parser.upperLimit)
    }
  }
}

File: prime-number/golang/pn.go
package main

import (
    "fmt"
    "flag"
)

type Number interface {
    int32 | int64
}

func isPrimeNumber[T Number](primeNumbers []T, n T) bool {
    for _, primeNum := range primeNumbers {
        if primeNum * primeNum > n {
            break
        }
        if (n % primeNum) == 0 {
            return false
        }
    }
    return true
}

func run[T Number](upperLimit int, showResult bool) {
    primeNumbers := []T{2}

    for n := 3; n < upperLimit; n+=2 {
        _n := T(n)
        if isPrimeNumber(primeNumbers, _n) {
            primeNumbers = append(primeNumbers, _n)
        }
    }

    fmt.Printf("Count: %d\n",  len(primeNumbers))
    if (showResult) {
        for _, primeNum := range primeNumbers {
            fmt.Printf("%d\n", primeNum)
        }
    }
}

func main() {
    var upperLimit int
    var varTypeInt64 bool
    var showResult bool
    flag.IntVar(&upperLimit, "u", 100, "Upper Limit")
    flag.BoolVar(&varTypeInt64, "int64", false, "use long")
    flag.BoolVar(&showResult, "s", false, "Show prime numbers")
    flag.Parse()

    fmt.Printf("Upper Limit: %d\n", upperLimit)
    if (varTypeInt64) {
        fmt.Println("VarType: int64")
        run[int64](upperLimit, showResult)
    } else {
        fmt.Println("VarType: int32")
        run[int32](upperLimit, showResult)
    }
}
